src/Microsoft.VisualStudio.SDK.Analyzers/VSSDK013ThreadAffinitizedMEFConstruction.cs

```diff
@@ -0,0 +1,185 @@
+
+        /// <inheritdoc />
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Descriptor);
+
+        private ImmutableArray<TypeMatchSpec> MembersRequiringMainThread { get; set; }
+
+        /// <inheritdoc />
+        public override void Initialize(AnalysisContext context)
+        {
+            if (context is null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze);
+
+            // Register for compilation first so that we only activate the analyzer for applicable compilations
+            context.RegisterCompilationStartAction(startCompilation =>
+            {
+                this.MembersRequiringMainThread = AdditionalFilesHelpers.GetMembersRequiringMainThread(startCompilation.Options, startCompilation.CancellationToken);
+                INamedTypeSymbol? importingConstructorAttribute = startCompilation.Compilation.GetTypeByMetadataName(Types.ImportingConstructorAttribute.FullName)?.OriginalDefinition;
+                INamedTypeSymbol? partImportsSatisfiedNotificationInterface = startCompilation.Compilation.GetTypeByMetadataName(Types.IPartImportsSatisfiedNotification.FullName)?.OriginalDefinition;
+                INamedTypeSymbol? exportAttributeType = startCompilation.Compilation.GetTypeByMetadataName(Types.ExportAttribute.FullName)?.OriginalDefinition;
+
+                var operationKinds = ImmutableArray.Create<OperationKind>(
+                    OperationKind.MethodReference,
+                    OperationKind.InstanceReference,
+                    OperationKind.FieldReference,
+                    OperationKind.ObjectOrCollectionInitializer,
+                    OperationKind.Invocation, // Method calls
+                    OperationKind.MemberInitializer, // For static member access
+                    OperationKind.PropertyReference, // For property access
+                    OperationKind.ObjectCreation); // For object creation scenarios
+
+                startCompilation.RegisterOperationAction(
+                    Utils.DebuggableWrapper(c => this.AnalyzeOperation(c, exportAttributeType, importingConstructorAttribute, partImportsSatisfiedNotificationInterface)),
+                    operationKinds);
+            });
+        }
+
+        private void AnalyzeOperation(
+            OperationAnalysisContext c,
+            INamedTypeSymbol? exportAttributeType,
+            INamedTypeSymbol? importingConstructorAttribute,
+            INamedTypeSymbol? partImportsSatisfiedNotificationInterface)
+        {
+            ISymbol containingSymbol = c.ContainingSymbol;
+            IOperation operation = c.Operation;
+            INamedTypeSymbol containingType = containingSymbol.ContainingType;
+
+            if (containingType is null || exportAttributeType is null)
```

Comment:
if you're going to bail on not having the exportAttributeType, it would be better to not have registered the operation in the first place in that case